internal/k8s/reconciler/listener.go

```diff
@@ -123,12 +129,118 @@ func (l *K8sListener) validateTLS(ctx context.Context) error {
 		}
 		l.status.ResolvedRefs.InvalidCertificateRef = certificateErr
 	} else {
-		l.certificates = []string{resource}
+		l.tls.Certificates = []string{resource}
+	}
+
+	if l.listener.TLS.Options != nil {
+		tlsMinVersion := l.listener.TLS.Options[tlsMinVersionAnnotationKey]
+		tlsMaxVersion := l.listener.TLS.Options[tlsMaxVersionAnnotationKey]
+		tlsCipherSuitesStr := l.listener.TLS.Options[tlsCipherSuitesAnnotationKey]
+
+		if tlsMinVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMinVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS min version")
+				return nil
+			}
+
+			if tlsCipherSuitesStr != "" {
+				if _, ok := tlsVersionsWithConfigurableCipherSuites[string(tlsMinVersion)]; !ok {
+					l.status.Ready.Invalid = errors.New("configuring TLS cipher suites is only supported for TLS 1.2 and earlier")
+					return nil
+				}
+			}
+
+			l.tls.MinVersion = string(tlsMinVersion)
+		}
+
+		if tlsMaxVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMaxVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS max version")
+				return nil
+			}
+
+			l.tls.MaxVersion = string(tlsMaxVersion)
+		}
+
+		if tlsCipherSuitesStr != "" {
+			// split comma delimited string into string array and trim whitespace
+			tlsCipherSuitesUntrimmed := strings.Split(string(tlsCipherSuitesStr), ",")
+			tlsCipherSuites := tlsCipherSuitesUntrimmed[:0]
+			for _, c := range tlsCipherSuitesUntrimmed {
+				tlsCipherSuites = append(tlsCipherSuites, strings.TrimSpace(c))
+			}
+
+			// validate each cipher suite in array
+			for _, c := range tlsCipherSuites {
+				if _, ok := supportedTlsCipherSuites[c]; !ok {
+					l.status.Ready.Invalid = fmt.Errorf("unrecognized or unsupported TLS cipher suite: %s", c)
+					return nil
+				}
+			}
+
+			// set cipher suites on listener TLS params
+			l.tls.CipherSuites = tlsCipherSuites
+		}
 	}
 
 	return nil
 }
 
+var supportedTlsVersions = map[string]struct{}{
+	"TLS_AUTO": {},
+	"TLSv1_0":  {},
+	"TLSv1_1":  {},
+	"TLSv1_2":  {},
+	"TLSv1_3":  {},
+}
+
+var tlsVersionsWithConfigurableCipherSuites = map[string]struct{}{
+	// Remove these two if Envoy ever sets TLS 1.3 as default minimum
+	"":         {},
+	"TLS_AUTO": {},
+
+	"TLSv1_0": {},
+	"TLSv1_1": {},
+	"TLSv1_2": {},
+}
+
+// TODO: move this somewhere where it can be imported as default config in
+// internal/adapters/consul when CipherSuites is undefined
+var defaultTlsCipherSuites = []string{
+	"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
+	"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
+	"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+	"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
+	"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
+	"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
+}
```

Comment:
Is there a common location this could live to be imported by both `k8s/reconciler` and `adapters/consul`?

Reply:
Generally it depends on how tied it is to either k8s or consul. If it's highly specific to `consul` because it maps to the config entry values, maybe the `consul` subpackage, otherwise things that are shared logic but not tied to any particular integration, but more just shared internally in our `consul-api-gateway` world, I've been putting under `common`.